### PR TITLE
Inject a custom DNS resolver

### PR DESCRIPTION
**Pull request by rzabini**
The proposed change addresses the following use case:

_As a developer, I want to redirect the browser to a local server, without changing the host name in the location, to test a local instance of a production website_.

This is usually achieved by changing the 'hosts' file on the local computer, however:
* the developer may not have the permission to change that file, or
* the developer may want to perform the redirection only on a single browser profile, while still being able to reach the real site in another browser profile.

In this change, the proxy uses a custom resolver, configured in `HttpProxyServerConfig`, which resolves every name to `localhost`.

### CODE REVIEW

**Comment by monkeyWie** [comment]
I think this is weighty, for this case, it is recommended to refer [InterceptForwardHttpProxyServer.java](https://github.com/monkeyWie/proxyee/blob/master/src/test/java/com/github/monkeywie/proxyee/InterceptForwardHttpProxyServer.java).

Of course, inject a custom DNS resolver is a nice feature, I will merge it.

File: pom.xml [hunk 1]
```diff
@@ -189,5 +189,38 @@
...
+                                <phase>package</phase>
+                                <goals>
+                                    <goal>single</goal>
+                                </goals>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
```
Comment (monkeyWie) [inline pom.xml:224]: This is a library project, these should be done by the user who used this library.

File: src/main/java/com/github/monkeywie/proxyee/LoopbackProxyServerApp.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        System.out.println("start loopback proxy server");
+        int port = 9999;
+        if (args.length > 0) {
+            port = Integer.parseInt(args[0]);
+        }
+        new HttpProxyServer()
+                .serverConfig(new HttpProxyServerConfig(LoopbackAddressResolverGroup.INSTANCE))
+                .start(port);
+    }
+}
```
Comment (monkeyWie) [inline src/main/java/com/github/monkeywie/proxyee/LoopbackProxyServerApp.java:18]: Same reason.

File: src/main/java/com/github/monkeywie/proxyee/resolver/LoopbackAddressResolverGroup.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        @Override
+        protected void doResolveAll(String inetHost, Promise<List<InetAddress>> promise) {
+            try {
+                promise.setSuccess(Arrays.asList(SocketUtils.allAddressesByName("localhost")));
+            } catch (UnknownHostException unknownHostException) {
+                promise.setFailure(unknownHostException);
+            }
+        }
+    }
+}
```
Comment (monkeyWie) [inline src/main/java/com/github/monkeywie/proxyee/resolver/LoopbackAddressResolverGroup.java:51]: Same reason.

File: pom.xml (hunk 1 above)
Comment (rzabini) [reply pom.xml:224]: I agree.

File: src/main/java/com/github/monkeywie/proxyee/LoopbackProxyServerApp.java (hunk 2 above)
Comment (rzabini) [reply src/main/java/com/github/monkeywie/proxyee/LoopbackProxyServerApp.java:18]: I agree

**Comment by rzabini** [comment]
> I think this is weighty, for this case, it is recommended to refer [InterceptForwardHttpProxyServer.java](https://github.com/monkeyWie/proxyee/blob/master/src/test/java/com/github/monkeywie/proxyee/InterceptForwardHttpProxyServer.java).

Yes, this was my first try. However I am not sure that it is the exact same use case: I need to change the IP address, without changing the Http host header of the request.
In the example you suggest, the browser asks for the _baidu_ site, and the proxy fetches a URL from the _taobao_ site, but in order to get a valid response the host header must be changed, am I right?

> 
> Of course, inject a custom DNS resolver is a nice feature, I will merge it.

Thank you

**Comment by monkeyWie** [comment]
> In the example you suggest, the browser asks for the baidu site, and the proxy fetches a URL from the taobao site, but in order to get a valid response the host header must be changed, am I right?
Yes, the request host header will be changed, so it should not be suitable for your case, please remove the reviewed code and then I will merge it.
Thank you~

**Pull request merged**